slackclient/slackrequest.py

```diff
@@ -74,18 +75,26 @@ def do(self, token, request="?", post_data=None, domain="slack.com", timeout=Non
         # use the 'file' argument to point to a File ID.
         post_data = post_data or {}
 
-        # Check for plural fields and convert them to comma-separated strings if needed
-        for field in {'channels', 'users', 'types'} & set(post_data.keys()):
-            if isinstance(post_data[field], list):
-                post_data[field] = ",".join(post_data[field])
-
         # Move singular file objects into `files`
         upload_requests = ['files.upload']
 
+        # Move file content into requests' `files` param
         files = None
         if request in upload_requests:
             files = {'file': post_data.pop('file')} if 'file' in post_data else None
 
+        # Check for plural fields and convert them to comma-separated strings if needed
+        for field in {'channels', 'users', 'types'} & set(post_data.keys()):
+            if isinstance(post_data[field], list):
+                post_data[field] = ",".join(post_data[field])
+
+        # Convert any params which aren't integer, string or bool to JSON
+        # Example: `attachments` is a dict
+        for k, v in six.iteritems(post_data):
+            if not isinstance(v, (six.string_types, six.integer_types, bool)):
+                post_data[k] = json.dumps(v)
```

Comment:
Optionally you might want to handle `TypeError` (I'm assuming it [raises similar errors as `dump` but I could be lying](https://docs.python.org/2/library/json.html#json.dump)

Reply:
The wrapper isn't smart enough yet to know what types to expect for which fields, just the approved types across all possible fields 😞